### PR TITLE
Use generators to allow multiple "results" from action

### PR DESCRIPTION
**Pull request by beberlei**
One problem with the automagic approach of converting arrays and `RedirectRoute` objects returned from a controller is, that you cannot add additional metadata to the response, or add other "sideeffect" output such as flash messages.

This PR introduces two new abstractions `ControllerResultConverter` or `ControllerYieldApplier` that allow the return of a generator from a controller.

The templating code is moved to an implementation called `ArrayToTemplateResponseConverter`. The `RedirectListener` code is then added to another `RedirectConverter`.

We then have a list of yield appliers, meaning they apply the yielded results to the response object created by either array to template or redirectroute to response converters. the following appliers are added:

- Symfony Cookie
- `QafooLabs\MVC\Headers`
- `QafooLabs\MVC\Flash`

The idea is to return multiple results from controllers that are combined together:

```php
use QafooLabs\MVC\Headers;
use QafooLabs\MVC\Flash;
use QafooLabs\MVC\RedirectRoute;
use Symfony\Component\HttpFoundation\Cookie;

class ArticlesController
{
    public function listAction()
    {
        yield new Headers(['X-Foo' => 'Bar']);
        return ['articles' => $this->repository->findAll()];
    }

    public function editAction(Article $article)
    {
        yield new Flash('info', 'Edited Article');
        yield new Cookie('foo', 'bar');

        return new RedirectRoute('list');
    }
}
```

What is missing? A way to easily manipulate the Cache API of an response. It has a lot of methods, so that is difficult. What would be the best abstraction? The most generic one would be a closure, that gets passed the response:

```php
yield function (Response $response) { $response->setPublic(true); };
```

But that is quite technical.

### CODE REVIEW

File: src/QafooLabs/Bundle/NoFrameworkBundle/Controller/ResultConverter/ArrayToTemplateResponseConverter.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    public function __construct(EngineInterface $templating, TemplateGuesser $guesser, string $engine)
+    {
+        $this->templating = $templating;
+        $this->guesser = $guesser;
+        $this->engine = $engine;
+    }
+
+    public function supports($result)
+    {
+        return is_array($result) || $templateView instanceof TemplateView;
```
Comment (tolry) [inline src/QafooLabs/Bundle/NoFrameworkBundle/Controller/ResultConverter/ArrayToTemplateResponseConverter.php]: wrong variable name `$templateView`, should be `$result`?

**Comment by beberlei** [comment]
@tolry the order does not matter unless you return an array twice for example, then the last one overwrites the newer one at the moment. The algorithm about what happens must be documented for sure.

**Comment by beberlei** [comment]
I modified the approach to require a return statement, which is converted to response. and then apply all yields to the response afterwards. This way there is no way to return multiple responses at the same time.

File: src/QafooLabs/Bundle/NoFrameworkBundle/EventListener/ViewListener.php [hunk 2]
```diff
@@ -57,35 +41,64 @@ public function onKernelView(GetResponseForControllerResultEvent $event)
...
+        if ($response) {
+            $event->setResponse($response);
+        }
+    }
+
+    private function unrollGenerator(Generator $generator, Request $request): ?Response
+    {
+        $yields = [];
+
+        foreach ($generator as $element) {
```
Comment (henrikbjorn) [inline src/QafooLabs/Bundle/NoFrameworkBundle/EventListener/ViewListener.php]: would `iterator_to_array` do the same?

**Comment by timglabisch** [comment]
i don't like the idea.

i've another vision how to use `yield` inside of controllers.
when using an async programming model most people try to use something like promises and futures.
I often use another approach i call `requirements`.

modern async frameworks often treat controllers as futures. we could do something similar in php, this would look more or less like 


```php

class ArticlesController
{
    public function listAction(): Future
    {
        return loadArticles().andThen(function($articles) {
             render('template.twig', $articles)
         })
    }
}
```

the code is fine, but has 2 main disadvantages.
1. it is not that readable
2. it doesn't help to parallelism is hard

instead sometimes i use a concept i call "requirement pattern".
here is a prototype for a controller that needs to render an template that has the "top Articles" and the Article with the ID 1.

 ```php
class ArticlesController
{
    public function listAction(): Requirement
    {
       [$articles, $article1] = yield [
          ArticlesRequirement::TopArticles(),
          ArticlesRequirement::WithID(1)
      ];

       return new TemplateRequirement('template.twig', $articles, article1);
    }
}
```

so what am i doing here? i am treating the controller as a `Requirement` that has to be resolved. Every Requirement could have sub-Requirements. In This case the Top-Articles and Article 1.

The Idea is to use `Yield` to specify something that is required later on.

the approach has one huge advantage  in contrast of Promises. Requirements are Typed. In a promise based apprach you would have to load the top articles and then the article with the id 1.
In a Requirement-Based approch you could Resolve the TopArticles to N  WithID Requirements, group them all together and load them in one single "in" query. This way you could load dynamic deep nested 
requirement graphs  in an async and readable way. I use this approach in production to resolve thousands of requirements in a graph to render "productlists" 

so now i tryed to explain how i would use "yield" in symfony controllers. 

i would solve the cookie example with a ParamConverter.

```php 
use Symfony\Component\HttpFoundation\Cookie;

class ArticlesController
{
    public function listAction(ResponseExtra $resp)
    {
        $resp->addHeader('X-Foo', 'Bar');

        return ['articles' => $this->repository->findAll()];
    }
```

**Comment by beberlei** [comment]
@timglabisch php is not async, it doesn't make sense to pretend it is. 

The requirement pattern as you call it is something bastian hoffmann has talked about in context of the research gate architecture. It is a great pattern, and generators make sense there, but its not really useful in regular MVC.

The `ResponseExtra` argument you proposed is not good style imho. The response is a return parameter, as are all its amendments.

**Comment by timglabisch** [comment]
>  but its not really useful in regular MVC.
it is interesting everywhere where "data" have to be loaded, async is not the only use case.

> The response is a return parameter, as are all its amendments.
after looking a while at your example i my mind changed a bit. May it's a good idea to treat a controller just as a generator of whatever (Cookies, Flashs, Requirements, Promises, ...).

File: README.md [hunk 3]
```diff
@@ -297,15 +326,9 @@ public function indexAction(Session $session)
 
 ## ParamConverter for Flash Messages
 
-You can pass a flash object as an argument to a controller:
-
-```
-use QafooLabs\MVC\Flash;
+Passing `QafooLabs\MVC\Flash` is not supported anymore. You must
+migrate the code to use `yield Flash($type, $message);` instead.
```
Comment (MetalArend) [inline README.md]:
```suggestion
Shouldn't this be `yield new Flash($type, $message);` instead?
```

File: src/QafooLabs/Bundle/NoFrameworkBundle/Controller/ResultConverter/ArrayToTemplateResponseConverter.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    private $engine;
+
+    public function __construct(EngineInterface $templating, TemplateGuesser $guesser, string $engine)
+    {
+        $this->templating = $templating;
+        $this->guesser = $guesser;
+        $this->engine = $engine;
+    }
+
+    public function supports($result)
```
Comment (localheinz) [inline src/QafooLabs/Bundle/NoFrameworkBundle/Controller/ResultConverter/ArrayToTemplateResponseConverter.php]: What do you think about adding a return type declaration here?

File: src/QafooLabs/Bundle/NoFrameworkBundle/Controller/ResultConverter/ArrayToTemplateResponseConverter.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        return $this->makeResponseFor(
+            $controller,
+            $result,
+            $request->getRequestFormat()
+        );
+    }
+
+    private function makeResponseFor($controller, TemplateView $templateView, $requestFormat)
```
Comment (localheinz) [inline src/QafooLabs/Bundle/NoFrameworkBundle/Controller/ResultConverter/ArrayToTemplateResponseConverter.php]: What do you think about adding a return type declaration here?

File: src/QafooLabs/Bundle/NoFrameworkBundle/Controller/ResultConverter/ControllerResultConverter.php [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+namespace QafooLabs\Bundle\NoFrameworkBundle\Controller\ResultConverter;
+
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\Request;
+
+interface ControllerResultConverter
+{
+    public function supports($result);
```
Comment (localheinz) [inline src/QafooLabs/Bundle/NoFrameworkBundle/Controller/ResultConverter/ControllerResultConverter.php]: What do you think about adding a return type declaration here?

File: src/QafooLabs/Bundle/NoFrameworkBundle/Controller/ResultConverter/ControllerYieldApplier.php [hunk 7]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+namespace QafooLabs\Bundle\NoFrameworkBundle\Controller\ResultConverter;
+
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\Request;
+
+interface ControllerYieldApplier
+{
+    public function supports($yield);
```
Comment (localheinz) [inline src/QafooLabs/Bundle/NoFrameworkBundle/Controller/ResultConverter/ControllerYieldApplier.php]: What do you think about adding a return type declaration here?

File: src/QafooLabs/Bundle/NoFrameworkBundle/EventListener/ViewListener.php [hunk 8]
```diff
@@ -5,47 +5,31 @@
...
-
-    /**
-     * @param string $engine
-     */
-    public function __construct(EngineInterface $templating, TemplateGuesser $guesser, $engine)
+    private $converters = [];
+
+    private $yieldAppliers = [];
+
+    public function addConverter(ControllerResultConverter $converter)
```
Comment (localheinz) [inline src/QafooLabs/Bundle/NoFrameworkBundle/EventListener/ViewListener.php]: What do you think about adding a return type declaration here?

File: src/QafooLabs/Bundle/NoFrameworkBundle/EventListener/ViewListener.php [hunk 9]
```diff
@@ -5,47 +5,31 @@
...
+    {
+        $this->converters[] = $converter;
+    }
+
+    public function addYieldApplier(ControllerYieldApplier $applier)
```
Comment (localheinz) [inline src/QafooLabs/Bundle/NoFrameworkBundle/EventListener/ViewListener.php]: What do you think about adding a return type declaration here?

File: src/QafooLabs/Bundle/NoFrameworkBundle/EventListener/ViewListener.php [hunk 10]
```diff
@@ -5,47 +5,31 @@
...
     {
-        $this->templating = $templating;
-        $this->guesser    = $guesser;
-        $this->engine     = $engine;
+        $this->yieldAppliers[] = $applier;
     }
 
     public function onKernelView(GetResponseForControllerResultEvent $event)
```
Comment (localheinz) [inline src/QafooLabs/Bundle/NoFrameworkBundle/EventListener/ViewListener.php]: What do you think about adding a return type declaration here?

File: src/QafooLabs/Bundle/NoFrameworkBundle/EventListener/ViewListener.php [hunk 11]
```diff
@@ -57,35 +41,60 @@ public function onKernelView(GetResponseForControllerResultEvent $event)
...
+        $response = ($result instanceof Generator)
+            ? $this->unrollGenerator($result, $request)
+            : $this->convert($result, $request);
+
+        if ($response) {
+            $event->setResponse($response);
         }
+    }
+
+    private function unrollGenerator(Generator $generator, Request $request): ?Response
```
Comment (localheinz) [inline src/QafooLabs/Bundle/NoFrameworkBundle/EventListener/ViewListener.php]: I think a nullable return type declaration is not needed here, what do you think?

File: src/QafooLabs/Bundle/NoFrameworkBundle/Tests/Controller/ResultConverter/ArrayToTemplateResponseConverterTest.php [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+use QafooLabs\Bundle\NoFrameworkBundle\View\TemplateGuesser;
+use QafooLabs\MVC\TemplateView;
+
+class ArrayToTemplateResponseConverterTest extends TestCase
+{
+    private $templating;
+    private $guesser;
+    private $converter;
+
+    public function setUp()
```
Comment (localheinz) [inline src/QafooLabs/Bundle/NoFrameworkBundle/Tests/Controller/ResultConverter/ArrayToTemplateResponseConverterTest.php]: What do you think about reducing the visibility here to `protected`? This is the visibility as declared on the parent class; there's probably no need to increase it.

File: src/QafooLabs/Bundle/NoFrameworkBundle/Tests/Controller/ResultConverter/ArrayToTemplateResponseConverterTest.php [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        $this->converter = new ArrayToTemplateResponseConverter(
+            $this->templating = \Phake::mock(EngineInterface::class),
+            $this->guesser = \Phake::mock(TemplateGuesser::class),
+            'twig'
+        );
+    }
+
+    public function testSupports()
```
Comment (localheinz) [inline src/QafooLabs/Bundle/NoFrameworkBundle/Tests/Controller/ResultConverter/ArrayToTemplateResponseConverterTest.php]: What do you think about adding a return type declaration here, and on all other test methods?

File: src/QafooLabs/Bundle/NoFrameworkBundle/Tests/Controller/ResultConverter/ArrayToTemplateResponseConverterTest.php [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+use PHPUnit\Framework\TestCase;
+
+use QafooLabs\Bundle\NoFrameworkBundle\Controller\ResultConverter\ArrayToTemplateResponseConverter;
+use Symfony\Component\Templating\EngineInterface;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\Request;
+use QafooLabs\Bundle\NoFrameworkBundle\View\TemplateGuesser;
+use QafooLabs\MVC\TemplateView;
+
+class ArrayToTemplateResponseConverterTest extends TestCase
```
Comment (localheinz) [inline src/QafooLabs/Bundle/NoFrameworkBundle/Tests/Controller/ResultConverter/ArrayToTemplateResponseConverterTest.php]: What do you think about marking all tests as `final`? There's probably no need for them to be not `final`.

**Pull request merged**